### PR TITLE
decoupling recorder, observer and pattern matcher

### PR DESCRIPTION
**Pull request by Sczlog**
refactor code structure for the sake of further development.
now recorder have three module:
1. observer:
A module spys on the user interaction, it will listen event and emit it to recorder.
2. matcher:
A module processs sending event and tranform them into a user action, then send it back to recorder.
3. recorder:
A module organizes observers and matcher, a recorder can have many observers and one matcher, it will provide a bridge between observers and matcher, and communicate with outer.

SimpleRecorder is an alternative to the previous recorder.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,130 +1,251 @@
...
-  constructor(options: RecorderOptions) {
-    const { win, doc, onEmit, metaQuerier } = options;
-    this.onEmit = onEmit;
-    this.eventObserver = new EventObserver(
-      win,
-      doc,
-      this.eventHandler.bind(this),
-    );
-    this.metaQuerier = metaQuerier || new BasicMetaQuerier();
+    this._recorder.extendAction<PatternMatcherExtendParams>({
```
Comment (netweng) [inline src/index.ts]: maybe we can have `buiid-in-action` files to manage these code

File: src/matcher.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    const { action } = this.matchPattern();
+    this.emitter?.emit(
+      MatcherKey.EMIT,
+      action,
+      this.currentEvents.splice(0, this.currentEvents.length),
+      this.currentTarget,
+    );
+  }
+
+  private handleNewEvent(stepEvent: StepEvent, target: HTMLElement | null) {
```
Comment (netweng) [inline src/matcher.ts:154]: which situation would be `continue`? which would be `return`?

File: src/recorder.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+  public extendAction<params extends IExtendParams>(action: params): IObserver {
+    const { observer } = action;
+    if (this.state === 'active') {
+      console.warn(
+        'cannot extend recorder when active, please suspend or stop recorder first',
+      );
+      return observer;
+    }
+    if (this.emitterMap.has(observer)) {
```
Comment (netweng) [inline src/recorder.ts]: we can have only one listener on an event ?

File: src/observers.ts [hunk 4]
```diff
@@ -62,12 +49,15 @@ export class EventObserver {
 
   private throttleManager = new ThrottleManager();
 
-  constructor(win: Window, doc: Document, onEmit: EmitHandler) {
+  constructor(win: Window, doc: Document) {
```
Comment (netweng) [inline src/observers.ts]: we still need doc?

Comment (Sczlog) [reply src/observers.ts]: We still need document element for scroll event,but we can use `window.document` to replace it.

File: src/matcher.ts (hunk 2 above)
Comment (Sczlog) [reply src/matcher.ts:154]: the handler should return `'return'|'emit'|'collect'|undefined`
`'return'` mean we should ignore this event and not process it.
`'undefined'` mean we should ignore current handler but keep processing it.
`'emit|collect'` mean we should do some action and stop current handler group and continue processing the event.

File: src/recorder.ts (hunk 3 above)
Comment (Sczlog) [reply src/recorder.ts]: It means one observer should only have a single group processing method.
When extended, they are immutable.
If you want to modify the action, you should firstly suspend the the recorder and remove this action and re-extend the observer.

File: src/matcher.ts (hunk 2 above)
Comment (netweng) [reply src/matcher.ts:154]: can we replace preStep, collectStep, postStep  by using:
``` typescript
shouldEmitBeforeCollectStep: () => boolean;
shouldCollect: () => boolean;
shouldEmitAfterCollectStep: () => boolean;
```

Comment (Sczlog) [reply src/matcher.ts:154]: I think boolean is not clear in this situation as the return value has 3 state, using enum or string will be better.
Name doesn't matter, I will change it.

Comment (netweng) [reply src/matcher.ts:154]: boolean | undefined ?

File: __tests__/units/matcher.spec.ts [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+  PatternMatcher,
+  PatternMatcherExtendParams,
+  MatcherKey,
+  StepEvent,
+  Step,
+} from '../../src/index';
+import { getMockedObserver } from '../util/mock';
+let matcher: PatternMatcher;
+
+const buildParams = (
```
Comment (netweng) [inline __tests__/units/matcher.spec.ts]: can we import type from  matcher.ts

File: __tests__/units/observer.spec.ts [hunk 6]
```diff
@@ -0,0 +1,604 @@
...
+      observer.start();
+      observer.suspend();
+      emitSpy = spyOn<any>(observer, 'onEmit');
+    });
+
+    afterEach(() => {
+      emitSpy.calls.reset();
+    });
+
+    describe('observe mouse interaction', () => {
```
Comment (netweng) [inline __tests__/units/observer.spec.ts]: should we need test every event handler?
maybe we just need to test the functions, but event observer should be test in puppeteer

**Comment by netweng** [comment]
LGTM

**Pull request merged**